# Lab Assignment One: Exploring Table Data 
**Christopher Cook, Bonita Davis, Anekah Kelley, Davis Lynn**

## 1. Business Understanding 

Test10101010

In [ ]:
import pandas as pd

## 2. Data Understanding

### 2.1 Data Description

In [8]:
import pandas as pd 

# find the data type
df = pd.read_csv('C:\CS1342\Programs\ML-Lab-1\ClassicHit.csv')

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15150 entries, 0 to 15149
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Track             15150 non-null  object 
 1   Artist            15150 non-null  object 
 2   Year              15150 non-null  int64  
 3   Duration          15150 non-null  int64  
 4   Time_Signature    15150 non-null  int64  
 5   Danceability      15150 non-null  float64
 6   Energy            15150 non-null  float64
 7   Key               15150 non-null  int64  
 8   Loudness          15150 non-null  float64
 9   Mode              15150 non-null  int64  
 10  Speechiness       15150 non-null  float64
 11  Acousticness      15150 non-null  float64
 12  Instrumentalness  15150 non-null  float64
 13  Liveness          15150 non-null  float64
 14  Valence           15150 non-null  float64
 15  Tempo             15150 non-null  float64
 16  Popularity        15150 non-null  int64 

Based on the dataframe information, there are no missing values in the dataset (there are 15150 rows in the data, and 15150 entries in each feature).

If there were missing values in the dataset, we could use mean or median imputation for numerical data (year, duration, danceability, energy, etc.), and mode for the categorical features (genre. etc.).

In [11]:
#data summary
df.describe()

Year      Duration  Time_Signature  Danceability        Energy  \
count  15150.000000  1.515000e+04    15150.000000  15150.000000  15150.000000   
mean    1984.709769  2.413028e+05        3.921650      0.574667      0.617691   
std       19.482460  1.057966e+05        0.348322      0.166208      0.239527   
min     1899.000000  2.300000e+04        0.000000      0.000000      0.000098   
25%     1971.000000  1.803700e+05        4.000000      0.461000      0.442000   
50%     1983.000000  2.228525e+05        4.000000      0.582000      0.644000   
75%     1998.000000  2.751530e+05        4.000000      0.696000      0.819000   
max     2024.000000  3.060650e+06        5.000000      0.988000      0.999000   

                Key      Loudness          Mode   Speechiness  Acousticness  \
count  15150.000000  15150.000000  15150.000000  15150.000000  15150.000000   
mean       5.257624     -9.154545      0.696634      0.075791      0.302651   
std        3.569713      4.289558      0.459727      0.077704      0.317793   
min        0.000000    -47.359000      0.000000      0.000000      0.000001   
25%        2.000000    -11.736750      0.000000      0.034325      0.024225   
50%        5.000000     -8.480500      1.000000      0.046400      0.167000   
75%        8.000000     -5.955500      1.000000      0.078400      0.557000   
max       11.000000      0.919000      1.000000      0.950000      0.996000   

       Instrumentalness      Liveness       Valence         Tempo  \
count      15150.000000  15150.000000  15150.000000  15150.000000   
mean           0.091046      0.200062      0.571146    120.664078   
std            0.229075      0.175196      0.250969     29.181107   
min            0.000000      0.012400      0.000000      0.000000   
25%            0.000000      0.091900      0.371000     98.672500   
50%            0.000084      0.130000      0.586000    118.777000   
75%            0.012400      0.257000      0.782000    138.073250   
max            0.987000      0.996000      0.989000    219.973000   

         Popularity  
count  15150.000000  
mean      43.025809  
std       20.788038  
min        0.000000  
25%       28.000000  
50%       44.000000  
75%       59.000000  
max       98.000000

In [12]:
data_des = pd.DataFrame()

data_des['Attributes'] = df.columns
data_des['Description'] = ['Name of song', 'Name of singer', 'Year released', 'Length of song in milliseconds', 'Musical time signature',
        'Suitability of dancing (0-1)', 'Intensity of the song (0-1)', 'Key of the track (0-11)', 'Loudness of track in dB', 'Modality (major or minor)',
        'Measures the presence of spoken words (0-1)', 'Measure of how acoustic the track is (0-1)', 'Measure of likelihood that track is instrumental (0-1)',
        'Measure of the presence of a live audience (0-1)', 'Measure of musical positiveness (0-1)', 'Speed of the track (BPM)', 'Track popularity (0-100)',
        'Musical genre of the song']
data_des['Scales'] = ['Nominal', 'Nominal', 'Ordinal', 'Ratio', 'Nominal', 'Ratio', 'Ratio', 'Ordinal', 'Ratio', 'Ordinal',
        'Ratio', 'Ratio', 'Ratio', 'Ratio', 'Ratio', 'Ratio', 'Ratio', 'Ordinal']
data_des['Discrete/Continuous'] = ['Discrete', 'Discrete', 'Discrete', 'Continuous', 'Discrete', 'Continuous', 'Continuous', 'Discrete',
        'Continuous', 'Discrete', 'Continuous', 'Continuous', 'Continuous', 'Continuous', 'Continuous', 'Continuous', 'Continuous', 'Discrete']
data_des['Range'] = ['N/A', 'N/A', '1899-2024', '23.0k-3.06m', '3 or 4', '0.0-1.0', '0.0-1.0', '0-11', '-47.4-0.92', '0 or 1', '0.0-1.0', '0.0-1.0', '0.0-1.0', 
        '0.0-1.0', '0.0-1.0', '0-220', '0-100', 'N/A']

data_des

Attributes                                        Description  \
0              Track                                       Name of song   
1             Artist                                     Name of singer   
2               Year                                      Year released   
3           Duration                     Length of song in milliseconds   
4     Time_Signature                             Musical time signature   
5       Danceability                       Suitability of dancing (0-1)   
6             Energy                        Intensity of the song (0-1)   
7                Key                            Key of the track (0-11)   
8           Loudness                            Loudness of track in dB   
9               Mode                          Modality (major or minor)   
10       Speechiness        Measures the presence of spoken words (0-1)   
11      Acousticness         Measure of how acoustic the track is (0-1)   
12  Instrumentalness  Measure of likelihood that track is instrument...   
13          Liveness   Measure of the presence of a live audience (0-1)   
14           Valence              Measure of musical positiveness (0-1)   
15             Tempo                           Speed of the track (BPM)   
16        Popularity                           Track popularity (0-100)   
17             Genre                          Musical genre of the song   

     Scales Discrete/Continuous        Range  
0   Nominal            Discrete          N/A  
1   Nominal            Discrete          N/A  
2   Ordinal            Discrete    1899-2024  
3     Ratio          Continuous  23.0k-3.06m  
4   Nominal            Discrete       3 or 4  
5     Ratio          Continuous      0.0-1.0  
6     Ratio          Continuous      0.0-1.0  
7   Ordinal            Discrete         0-11  
8     Ratio          Continuous   -47.4-0.92  
9   Ordinal            Discrete       0 or 1  
10    Ratio          Continuous      0.0-1.0  
11    Ratio          Continuous      0.0-1.0  
12    Ratio          Continuous      0.0-1.0  
13    Ratio          Continuous      0.0-1.0  
14    Ratio          Continuous      0.0-1.0  
15    Ratio          Continuous        0-220  
16    Ratio          Continuous        0-100  
17  Ordinal            Discrete          N/A

This table shows the attributes' description, scales, and range.

### 2.2 Attributes
There are 18 different attributes collected in the dataset, so we focused on the attributes that we used for our visualization.

1)  **Year:** This is the year that the track was released. The years range from 1899 to 2024, with the most tracks being from 1974-1980, followed by 1980-1986.
2)  **Key:** Describes the key of the track -- the conversion of which is: *0 = C, 1 = C#, 2 = D, 3 = Ef, 4 = E, 5 = F, 6 = F#, 7 = G, 8 = Af, 9 = A, 10 = Bf, 11 = B*. C and G were the most common keys.
3)  **Mode:** Shows if the key is major (1) or minor (0). The vast majority of songs were in a major key.
4)  **Popularity:** Measures the track's popularity on a scale of 0-100.  The popularity is essentially a bell curve, skewed slightly to the right.
5)  **Genre:** Describes the genre of the song. Pop, metal, country, R&B, and Rock are some of the most common.

### 2.3 Data Quality

As previously stated, there are no missing values in the dataset.

We should check whether there are duplicate values:

In [15]:
#check for duplicates
idx = df.duplicated()
len(df[idx])

1

This means that there is one duplicated song. We should see which it is:

In [18]:
#check which is a duplicate
df[idx]

Track Artist  Year  Duration  Time_Signature  Danceability  Energy  \
3640  Happy Now   Zedd  2019    207029               3         0.503   0.741   

      Key  Loudness  Mode  Speechiness  Acousticness  Instrumentalness  \
3640    4    -5.128     1        0.047         0.579          0.000001   

      Liveness  Valence   Tempo  Popularity Genre  
3640    0.0712    0.458  88.335          63   EDM

This is a complete duplicate -- "Happy Now" by Zedd, released in 2019.  After verifying that it was a perfect duplicate, our group chose to delete the song from our dataset. This completed the cleaning of our data, as we ensured that there was no missing data and there are now no duplicates.

### 2.4 Data Splits

We created a new division in the data called "era", which is described as year values < 1982 and year values > 1982. This division helps us visualize how the music changed over time in an easily digestible manner. It also marks the beginning of widespread use of synthesizers in music.

## 3. Data Visualization

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
import matplotlib.gridspec as gridspec

print('Pandas:', pd.__version__)
print('Numpy:',np.__version__)

In [ ]:
df = pd.read_csv('ClassicHit.csv')

df.head()

In [ ]:
print(df.dtypes)

In [ ]:
# Create the JointGrid object
g = sns.JointGrid(data=df, x="Popularity", y="Danceability", space=0)

# Plot the 2D histogram with KDE on the margins
g.plot_marginals(sns.histplot, kde=True, bins=10)

# Set the aspect ratio of the x-axis marginal plot
#g.ax_marg_x.set_aspect(1.0 / g.ax_marg_x.get_data_ratio(), adjustable='box')

# Add the 2D histogram to the central plot
ax = g.ax_joint
sns.histplot(x='Popularity', y='Valence', data=df, bins=10, ax=ax)

# Now add text annotations to the histogram cells
coords = ax.collections[0].get_coordinates()
half_width = (coords[0, 1, 0] - coords[0, 0, 0]) / 2
half_height = (coords[1, 0, 1] - coords[0, 0, 1]) / 2

for v, (xv, yv) in zip(ax.collections[0].get_array().ravel(), coords[:-1, :-1, :].reshape(-1, 2)):
    if not np.ma.is_masked(v):
        ax.text(xv + half_width, yv + half_height, f'{v:.0f}', ha='center', va='center', color='white')

# Display the plot
plt.show()

In [ ]:
# Start by just plotting what we previously grouped!
plt.style.use('ggplot')

fig = plt.figure(figsize=(15,5))

plt.subplot(1,3,1)
df.Popularity.plot.hist(bins=20)
plt.title('Histogram, 20 bins')

plt.subplot(1,3,2)
df.Popularity.plot.kde(bw_method=0.2)
plt.title('Kernel Density, 0.2 bandwidth')

plt.subplot(1,3,3)
df.Popularity.plot.hist(bins=20)
df.Popularity.plot.kde(bw_method=0.1, secondary_y=True)
plt.title('Hist and KDE, 20 bins and 0.1 bw')

plt.show()

In [ ]:
import warnings
warnings.simplefilter('ignore', DeprecationWarning)
%matplotlib inline 

# External package: conda install missingno 
import missingno as mn

mn.matrix(df)
plt.title("Not Sorted",fontsize=22)

plt.show()

In [ ]:
# let's break up the age variable into a discrete new variable
df['Era'] = pd.cut(df['Year'],
                            [0,1985,2027],
                            labels=['Before1985','After1985']) # this creates a new variable
df.Era.describe()

In [ ]:
# first group the data
df_grouped = df.groupby(by=['Genre','Era'])

# tabulate survival rates of each group
Popularity_rate = df_grouped.Popularity.sum() / df_grouped.Popularity.count()


plt.figure(figsize=(18, 12))
# show in a bar chart using builtin pandas API
ax = Popularity_rate.plot(kind='barh')
plt.title('Popularity Percentages by Genre and Era')
plt.show()

In [ ]:
# the cross tab operator provides an easy way to get these numbers
survival = pd.crosstab([df['Genre'],
                        df['Era']], # categories to cross tabulate
                       df.Mode.astype(bool)) # how to group
plt.figure(figsize=(18, 12))
survival.plot(kind='barh', stacked=True)
plt.show()

In [ ]:
# plot overall cross tab with both groups
plt.figure(figsize=(18, 10))
ax1 = plt.subplot(1,3,1)
ax2 = plt.subplot(1,3,2)
ax3 = plt.subplot(1,3,3)

pd.crosstab([df['Era']], # categories to cross tabulate
            df.Mode.astype(bool)).plot(kind='bar', stacked=True, ax=ax1) 

pd.crosstab([df['Genre']], # categories to cross tabulate
            df.Mode.astype(bool)).plot(kind='bar', stacked=True, ax=ax2) 

pd.crosstab([df['Key']], # categories to cross tabulate
            df.Mode.astype(bool)).plot(kind='bar', stacked=True, ax=ax3) 

plt.show()

#FUCK FUCK FUCK

In [ ]:
# plot the correlation matrix 
vars_to_use = ['Year', 'Key', 'Duration', 'Tempo', 'Popularity', 'Energy', 'Time_Signature'] # pick vars
corr_matrix = df[vars_to_use].corr()

plt.figure(figsize=(18,10))

plt.pcolor(df[vars_to_use].corr()) # do the feature correlation plot

# fill in the indices
plt.yticks(np.arange(0.5, len(vars_to_use), 1), vars_to_use)
plt.xticks(np.arange(0.5, len(vars_to_use), 1), vars_to_use)
plt.colorbar()

# Add ticks for variables
plt.yticks(np.arange(0.5, len(vars_to_use), 1), vars_to_use)
plt.xticks(np.arange(0.5, len(vars_to_use), 1), vars_to_use)

# Annotate each cell with the correlation coefficient
for i in range(len(vars_to_use)):
    for j in range(len(vars_to_use)):
        plt.text(j + 0.5, i + 0.5, f'{corr_matrix.iloc[i, j]:.2f}', 
                 ha='center', va='center', color='black')
        
plt.show()

In [ ]:
# sns boxplot
plt.subplots(figsize=(20, 5))

plt.subplot(1,3,1)
sns.boxplot(x="Mode", y="Popularity", hue="Era", data=df)
plt.title('Boxplot Example')

plt.subplot(1,3,2)
sns.violinplot(x="Mode", y="Popularity", hue="Era", data=df)
plt.title('Violin Example')

plt.subplot(1,3,3)
sns.swarmplot(x="Mode", y="Popularity", hue="Era", data=df, s=3) # s controls marker size (like bins or bw)
plt.title('Swarm Example')

plt.show()

In [ ]:
sns.violinplot(x="Mode", y="Popularity", hue="Era", data=df, 
               split=True, # split across violins
               inner="quart") # show innner stats like mean, IQR, 
               
plt.show()

## 4. Dimensionality Reduction